connman/src/nat.c

```diff
@@ -193,26 +191,83 @@ static void free_nat(struct connman_nat *nat)
 	g_free(nat);
 }
 
+static int count_nat(int family)
+{
+	GHashTableIter iter;
+	gpointer key;
+	gpointer value;
+	int count = 0;
+
+	DBG("family %d", family);
+
+	if (!nat_hash || g_hash_table_size(nat_hash) == 0)
+		return 0;
+
+	g_hash_table_iter_init(&iter, nat_hash);
+	while (g_hash_table_iter_next(&iter, &key, &value)) {
+		struct connman_nat *nat = value;
+
+		if (nat->family == family)
+			count++;
+	}
+
+	return count;
+}
+
+/* Enable is called before adding and disable after removal of nat */
+static int try_ip_forward(int family, bool enable)
+{
+	char v;
+	int count;
+	int err;
+
+	switch (family) {
+	case AF_INET:
+		v = '4';
+		break;
+	case AF_INET6:
+		v = '6';
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	count = count_nat(family);
+
+	/* First nat enables */
+	if (enable && count > 0) {
+		DBG("IPv%c forward already enabled, nat count %d", v, count);
+		return -EALREADY;
+	/* Last nat disables */
+	} else if (!enable && count != 0) {
```

Comment:
In case we are trying to enable it is good to report that it has been already enabled. In case of disable I think it should not be an error if there is still more nat's present. This is called before adding and after removing. I wouldn't dare to touch this anymore.